contracts/index_executor/src/task.rs

```diff
@@ -272,221 +420,396 @@ impl Task {
-                transfer_step.amount = settle_balance.min(transfer_step.flow);
-
-                // sender's balance
-                let latest_b0 = worker_account.get_balance(
-                    transfer_step.chain.clone(),
-                    transfer_step.asset.clone(),
-                    context,
-                )?;
-                // recipeint's balance
-                let latest_b1 = worker_account.get_balance(
-                    transfer_step.chain.clone(),
-                    transfer_step.asset.clone(),
-                    context,
-                )?;
-                transfer_step.b0 = Some(latest_b0);
-                transfer_step.b1 = Some(latest_b1);
+        let chain = context
+            .registry
+            .get_chain(&self.source)
+            .map(Ok)
+            .unwrap_or(Err("MissingChain"))?;
+        let account = match chain.chain_type {
+            ChainType::Evm => worker_account.account20.to_vec(),
+            ChainType::Sub => worker_account.account32.to_vec(),
+        };
+        let claim_nonce = self.claim_nonce.ok_or("MissingClaimNonce")?;
+
+        // Check if already claimed success
+        let onchain_nonce = worker_account.get_nonce(&self.source, context)?;
+        if onchain_nonce > claim_nonce {
+            if tx::has_confirmed(&chain.tx_indexer_url, &account, claim_nonce)? {
+                Ok(true)
+            } else {
+                Err("ClaimFailed")
             }
-            _ => return Err("UnexpectedStep"),
+        } else {
+            Ok(false)
         }
-        Ok(())
+    }
+
+    fn claim_evm_actived_tasks(
+        &mut self,
+        chain: Chain,
+        task_id: TaskId,
+        context: &Context,
+        nonce: u64,
+    ) -> Result<Vec<u8>, &'static str> {
+        let handler: H160 = H160::from_slice(&chain.handler_contract);
+        let transport = Eth::new(PinkHttp::new(chain.endpoint));
+        let handler = Contract::from_json(transport, handler, include_bytes!("./abi/handler.json"))
+            .map_err(|_| "ConstructContractFailed")?;
+        let worker = KeyPair::from(context.signer);
+
+        // We call claimAndBatchCall so that first step will be executed along with the claim operation
+        let first_step = &mut self.merged_steps[0];
+        first_step.set_spend(self.amount);
+        first_step.sync_origin_balance(context)?;
+        let params = (task_id, first_step.derive_calls(context)?);
+        pink_extension::info!("claimAndBatchCall params {:?}", &params);
+
+        // Estiamte gas before submission
+        let gas = resolve_ready(handler.estimate_gas(
+            "claimAndBatchCall",
+            params.clone(),
+            worker.address(),
+            Options::default(),
+        ))
+        .map_err(|e| {
+            pink_extension::error!(
+                "claimAndBatchCall: failed to estimate gas cost with error {:?}",
+                &e
+            );
+            "GasEstimateFailed"
+        })?;
```

Comment:
Nit: It's good to use Kevin's`ResultExt` trait like this:

```rust
use pink::ResultExt;

let gas = resolve_ready(...)
    .log_err("claimAndBatchCall: failed to estimate gas cost with error")
    .or(Err("GasEstimateFailed"))?;
```

Reply:
Done